robottelo/api/utils.py

```diff
@@ -661,3 +661,27 @@ def configure_puppet_test():
         'ak_name': ak_name,
         'env_name': env.name,
     }
+
+
+def wait_for_tasks(query, search_rate=1, max_tries=10):
+    """Search for tasks by specified search query and poll them to ensure that
+    task has finished.
+
+    :param query: Search query that will be passed to API call.
+    :param search_rate: Delay between searches.
+    :param max_tries: How many times search should be executed.
+    :raises: ``AssertionError`. If not tasks were found until timeout.
+    :return: List of ``nailgun.entities.ForemanTasks`` entities.
+    """
+    for _ in range(max_tries):
+        tasks = entities.ForemanTask().search(query={'search': query})
```

Comment:
I think this query already contains info about the status of the individual tasks, e.g.

```json
{
  "total": 558,
  "subtotal": 20,
  "page": 1,
  "per_page": 20,
  "sort": {
    "by": "started_at",
    "order": "DESC"
  },
  "results": [
    {
      "id": "bbb01263-f381-4df5-a64d-cba9f3b15442",
      "label": "Actions::Insights::EmailPoller",
      "pending": true,
      "username": null,
      "started_at": null,
      "ended_at": null,
      "state": "scheduled",
      "result": "pending",
      "progress": 0,
      "input": {},
      "output": {},
      "humanized": {
        "action": "Insights Email Notifications",
        "input": "N/A",
        "output": "N/A",
        "errors": "N/A"
      },
      "cli_example": null
    },
    {
      "id": "8f00d4ff-5d63-4088-8e25-4a85329cd174",
      "label": "Actions::Katello::Host::Unregister",
      "pending": false,
      "username": "foreman_admin",
      "started_at": "2017-09-12T14:29:33.000+02:00",
      "ended_at": "2017-09-12T14:29:34.000+02:00",
      "state": "stopped",
      "result": "success",
      "progress": 1,
      "input": {
        "remote_user": "admin",
        "remote_cp_user": "admin"
      },
      "output": {},
      "humanized": {
        "action": "Unregister Host",
        "input": [],
        "output": "",
        "errors": []
      },
      "cli_example": null
    }
  ]
}
```

Can we use this information to refine the list of the tasks that need checking? It would save us some additional polling and allow us to return || fail quicker

Reply:
^ not a blocker

Reply:
Discussed it with @rplevka on irc, pros and cons:

* Option 1: use as it is and let the nailgun do the magic
  * Pros: no additional code is required
  * Cons: In case of finished tasks it will perform additional GET request for each found task just to check that it has status 'stopped' and result 'success'

* Option 2: add additional boilerplate code
  * Pros: No additional requests
  * Cons: Need to re-implement status checks, exceptions raising that is done by nailgun.